### DISCUSSION TITLE
A puzzle about the INSADMass kernel

### DISCUSSION THREAD

**Discussion by Always-kimi** [idaholab/moose#27719] [Q&A Modules: Navier-Stokes] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A Navier Stokes is the most appropriate category for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (avoid screenshots if possible, triple back quotes before/after pasted text, etc)

### Question

Hi there,
I have a puzzle about the INSADMass kernel. Why does the INSADMass kernel need the variable pressure and not the variable velocity. 
```
  [mass]
    type = INSADMass
    variable = p
    use_displaced_mesh = true
  []
  [mass_pspg]
    type = INSADMassPSPG
    variable = p
    use_displaced_mesh = true
  []
```
I checked the source file for the INSAMass kernel and it seems call the INSADMaterial. The latter needs two variables which are velocity and pressure. 
```
registerMooseObject("NavierStokesApp", INSADMaterial);

InputParameters
INSADMaterial::validParams()
{
  InputParameters params = Material::validParams();
  params.addClassDescription("This is the material class used to compute some of the strong "
                             "residuals for the INS equations.");
  params.addRequiredCoupledVar("velocity", "The velocity");
  params.addRequiredCoupledVar(NS::pressure, "The pressure");
  params.addParam<MaterialPropertyName>("mu_name", "mu", "The name of the dynamic viscosity");
  params.addParam<MaterialPropertyName>("rho_name", "rho", "The name of the density");
  return params;
}
```
I feel confused why the variable in the continuity equation is pressure.

**Comment by grmnptr** [comment, marked as answer]
You can write the incompressible Navier-Stokes equations in a matrix notation (with operators as the blocks):
```
| A    B | |u| = |f1|
| B^T  0 | |p|   |0| 
```
$B$ is the gradient operator and $B^T$ is the divergence operator. In moose, the `variable` parameter in the kernel definition will determine where the residual entries go in the nonlinear system. You see by (just doing R=matrix*sol-rhs for example) that the residual from $B^T u$ will actually go to the parts of the residual vector which are associated with the pressure degrees of freedom. This is actually the reason why it is hard to solve the Navier-Stokes equations, you have a 0 pressure diagonal and you are forced to solve it by using either a segregated approach or a Schurr-complement-based approach.

**Reply by Always-kimi** [reply]
Oh, I got it. Thank you very much!
